HGCalAnalysis/plugins/HGCalAnalysis.cc

```diff
@@ -737,121 +856,123 @@ HGCalAnalysis::analyze(const edm::Event& iEvent, const edm::EventSetup& iSetup)
+	                    myTrack.vertex().position(),charge,propCoords);
+
+	            xp.push_back(propCoords.x);
+	            yp.push_back(propCoords.y);
+	            zp.push_back(propCoords.z);
+	        }
+	    }
+	    else
+	    {
+	        vtx = myTrack.endVertex().position();
+	    }
+	    auto orig_vtx=myTrack.vertex().position();
+
+	    allselectedgentracks.push_back(&mySimEvent->track(i));
+	    // fill branches
+	    genpart_eta.push_back(myTrack.momentum().eta());
+	    genpart_phi.push_back(myTrack.momentum().phi());
+	    genpart_pt.push_back(myTrack.momentum().pt());
+	    genpart_energy.push_back(myTrack.momentum().energy());
+	    genpart_dvx.push_back(vtx.x());
+	    genpart_dvy.push_back(vtx.y());
+	    genpart_dvz.push_back(vtx.z());
+
+	    genpart_ovx.push_back(orig_vtx.x());
+	    genpart_ovy.push_back(orig_vtx.y());
+	    genpart_ovz.push_back(orig_vtx.z());
+
+	    HGCal_helpers::coordinates hitsHGCal;
+	    toHGCalPropagator.propagate(myTrack.momentum(),
+	            orig_vtx,
+	            myTrack.charge(),hitsHGCal );
+
+
+	    genpart_exphi.push_back(hitsHGCal.phi);
+	    genpart_exeta.push_back(hitsHGCal.eta);
+	    genpart_exx.push_back(hitsHGCal.x);
+	    genpart_exy.push_back(hitsHGCal.y);
+
+	    genpart_fbrem.push_back(fbrem);
+	    genpart_pid.push_back(myTrack.type());
+	    genpart_gen.push_back(myTrack.genpartIndex());
+	    genpart_reachedEE.push_back(reachedEE);
+	    genpart_fromBeamPipe.push_back(true);
+
+	    genpart_posx.push_back(xp);
+	    genpart_posy.push_back(yp);
+	    genpart_posz.push_back(zp);
+	}
 
-	if( !readOfficialReco) {
-		unsigned int npart = part->size();
-		for(unsigned int i=0; i<npart; ++i) {
-
-			// event=0 is the hard scattering (all PU have event()>=1)
-			// bunchCrossing == 0 intime, buncCrossing!=0 offtime, standard generation has [-12,+3]
-			if((*part)[i].eventId().event() ==0 and (*part)[i].eventId().bunchCrossing()==0) {
-
-				// look for the generator particle, set to -1 for Geant produced paticles
-				int tp_genpart=-1;
-				if (!(*part)[i].genParticles().empty()) tp_genpart=(*part)[i].genParticle_begin().key();
-
-				// default values for decay position is outside detector, i.e. ~stable
-				int reachedEE=1;
-				double fbrem=-1;
-				float dvx=999.;
-				float dvy=999.;
-				float dvz=999.;
-				if((*part)[i].decayVertices().size()>=1) { //they can be delta rays, in this case you have multiple decay verices
-					dvx=(*part)[i].decayVertices()[0]->position().x();
-					dvy=(*part)[i].decayVertices()[0]->position().y();
-					dvz=(*part)[i].decayVertices()[0]->position().z();
-					if ((*part)[i].decayVertices()[0]->inVolume()) reachedEE=0; //if it decays inside the tracker volume
-				}
+	//associate gen particles to mothers
+	genpart_mother.resize(genpart_posz.size(),-1);
+	for(size_t i=0;i<allselectedgentracks.size();i++){
+	    auto track=allselectedgentracks.at(i);
+	    if(track->noMother())continue;
+	    for(size_t j=0;j<allselectedgentracks.size();j++){
```

Comment:
Can't we do the usual half matrix instead of the full N^2 loop?, Like starting j=i+1?

Reply:
The problem is that 'i' can be mother of 'j', but 'j' not mother of 'i'. We could ask the cross reference, that is true, then we could use the half matrix.

Reply:
@rovere Let me know if you're OK with leaving it like that or prefer the cross reference.

Reply:
@jkiesele maybe running on half the matrix and testing both cases (i is a mother for j and j is a mother for i) would cover all possibilities, moving the ifs inside the nested loops.

Reply:
Ok, I will implement it.
Can we move forward after I implemented it? 
It is already in the pipeline for quite some time.

Reply:
Sure, thanks, I had hoped that Luca @malgeri would comment on the other item. Maybe I'll catch him tomorrow.